### ISSUE TITLE
SystemOutOfMemory Exception when saving large Ultrasound DICOM files

### ISSUE THREAD

**Issue by FiifiB** [fo-dicom/fo-dicom#363]
Hi so i've been trying to edit each frame in a large Ultrasound DICOM file and save it to back to the hard drive. But i consistently get a SystemOutOfMemory Exception everytime i try and save the file with DicomFile.save() method. My code is below. My Ultrasound image is about 700MB but i'll need to edit larger Ultrasound Dicom Files so would be really helpful if anyone could assist in this. Thanks

```
//get first frame size            
DicomDataset newDataset = new DicomDataset();
dcmImage.Dataset.CopyTo(newDataset);
DicomPixelData pixelData = DicomPixelData.Create(newDataset, true);
//edits each frame in the DICOM image object
            for (int i = 1; i < dcmImage.NumberOfFrames; i++)
            {                
                int rows, columns;
                var frame = RenderBitmapSourceToBitmap(dcmImage.RenderImage(i).AsWriteableBitmap());
                frame = GetValidImage(frame);

                //Draws each masked region on the current frame
                foreach (Rectangle maskRegion in maskRegions)
                {
                    Point position = new Point(Canvas.GetLeft(maskRegion), Canvas.GetTop(maskRegion));
                    frame = drawMaskOnImage(maskRegion, position, frame, imgCtrl);
                }

                // converts frame to byte array and writes it back to PixelData stream object
                byte[] pixels = GetPixels(frame, out rows, out columns);
                newDataset.Add(DicomTag.Rows, (ushort)rows);
                newDataset.Add(DicomTag.Columns, (ushort)columns);
                System.Drawing.ImageConverter converter = new System.Drawing.ImageConverter();
                MemoryByteBuffer buffer = new MemoryByteBuffer(pixels);
                pixelData.AddFrame(buffer);

                GC.Collect();
            }
DicomFile dicomfile = new DicomFile(newDataset);
dicomfile.Save(outputFolder + @"\" + outputFileName + ".dcm");

```
### fo-dicom version and OS/platform

Latest Version from NuGet on Windows 7

**Comment by anders9ustafsson** [comment]
@FiifiB Which architecture are you running on? x64? x86?

Latest version, is that latest official (2.0.2) or pre-release (3.0.0-beta4)?

You are calling several of your own methods in the code above. Have you verified that those methods are not in any way leaking memory? Is it possible to dispose some of the allocated resources intermediately?

**Comment by FiifiB** [comment]
running on x64 and 2.0.2 of fo-dicom. The methods that are being called are fine. It seems that the majority of memory being used is by the byte buffer. But the program doesnt crash until the DicomFile object tries to save the file

**Comment by dhanzhang** [comment]
@anders9ustafsson   DicomFile.Save()  raised   OutOfMemoryException  when load large file, but that is not common

**Comment by FiifiB** [comment]
Any updates on this ?

**Comment by hdesouky** [comment]
Can you shed more light on this
Why are you using MemoryByeBuffer?
Did you tried FileByteBuffer?

**Comment by FiifiB** [comment]
Well im using MemoryBytebuffer because it takes a pixel array as a parameter. I need this because i edit the the pixels of each frame in a specific Dicom Image. FileByteBuffer doesnt do this. It takes a dicom file as a parameter instead. If you could please give an example of how you would use a FileByteBuffer to change the pixel data of a dicom file i would really appreciate that. Thank you

**Comment by FiifiB** [comment]
My Colleague has figured out how to use a FileByteBuffer to add frames to the PixelData object by using the FileByteSource class. But we are still running into the SystemOutOfMemory Exception when trying to save the DicomFile object to File with the DicomFile.save() method. Any suggestions anyone?

**Comment by anders9ustafsson** [comment]
@FiifiB How much RAM do you have on your system? Are you able to save the file on a computer with more memory? Are you able to upload an anonymized copy of the file somewhere for our testing purposes?

**Comment by FiifiB** [comment]
Thank you for the reply. I havent been able to save the file on any other computer. I've tried saving with a workstation laptop with 16GB of RAM and tried with Desktop Workstation with 64GB of RAM. Still no luck. I've been doing some snooping around the code. Particularly the DicomWriter class and the FileByteTarget class. I think the problem might be the BinaryStream used to write the DICOM file never gets flushed and so increases in size. And due the 2GB (in reality 1.3GB) limit of a CLR object, the OutofMemory exceptions keeps being thrown. I have tried with smaller data sets and the work fine. Unfortunately i cant share my data sets with you at the moment as they are not annonymized and are property of the company i work for. I hope this gives you some indication on where to look for a solution.

Thanks

**Comment by FiifiB** [comment]
So i now have a dataset i can share with you for testing. Please could you give me your email so i can send you a link to the dataset 

Thanks

**Comment by FiifiB** [comment]
So i think i have found the exact problem. The problem is in the CompositeByteBuffer class Data property

 public byte[] Data
        {
            get
            {
                byte[] data = new byte[Size];
                int offset = 0;
                foreach (IByteBuffer buffer in Buffers)
                {
                    System.Buffer.BlockCopy(buffer.Data, 0, data, offset, (int)buffer.Size);
                    offset += (int)buffer.Size;
                }
                return data;
            }
        }

when creating the byte array, it tries to create one big byte array to return to the DicomWriter and that is problem when it defines the length of the array as 1 580 000 000 which is the total size of the pixeldata of my DICOM file. This causes the SystemOutOfMemory Exception.

So instead i have cloned the repository and added a method to the IByteBuffer interface. The implementation of the method is below 

public byte[] GetDataAt(int index)
        {
            return Buffers[index].Data;
        }

Then in the DicomWriter when the element to write is the PixelData enumerate through the Buffers list in the CompositeByteBuffer and write each of the byte arrays. Below is what i have changed the current onElement method in the DicomWriter class to.


 public bool OnElement(DicomElement element)
        {
            IByteBuffer buffer = element.Buffer;
            WriteTagHeader(element.Tag, element.ValueRepresentation, element.Length);

            if (element.Tag == DicomTag.PixelData)
            {
                CompositeByteBuffer cbb = (CompositeByteBuffer)buffer;
                int countOfBuffers = cbb.Buffers.Count;
                int offset = 0;

                for (int i = 0; i < countOfBuffers; i++)
                {
                    var bytearr = cbb.GetDataAt(i);
                    _target.Write(bytearr, 0, (uint)bytearr.Length);
                    offset += bytearr.Length;
                }

                return true;
            }            

            
            if (buffer is EndianByteBuffer)
            {
                EndianByteBuffer ebb = buffer as EndianByteBuffer;
                if (ebb.Endian != Endian.LocalMachine && ebb.Endian == _target.Endian) buffer = ebb.Internal;
            }
            else if (_target.Endian != Endian.LocalMachine)
            {
                if (element.ValueRepresentation.UnitSize > 1) buffer = new SwapByteBuffer(buffer, element.ValueRepresentation.UnitSize);
            }

            _target.Write(buffer.Data, 0, buffer.Size);

            return true;
        }


Hope this helps anyone else

**Comment by anders9ustafsson** [comment]
@FiifiB Sorry for the late response, and many thanks for identifying a solution to the problem. Would you consider your solution general enough to include in the repository, or is it specific to your issue only?

If you still want to share the dataset, you may contact me directly at "info at cureos dot com".

**Comment by FiifiB** [comment]
@anders9ustafsson  Yes that solution is quite general. Have tested it on a number of DICOM images with different modalities. Feel free to test it out yourself. 

Will talk with my colleague about sharing our dataset and whether we need any further assistance. Thank you for your help though.

**Comment by anders9ustafsson** [comment]
@FiifiB I have chosen a slightly different approach from yours, but there is now a proposed fix of this issue available in PR #384. Please have a look to see if you find my solution feasible for your needs.
